### PR TITLE
[INLONG-8948][Sort] Add Redis connector on flink 1.15

### LINKED ISSUE #8948: [Feature][Sort]  Add Redis connector on flink 1.15
### Description

Parent issue: https://github.com/apache/inlong/issues/8336

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### LINKED ISSUE #9835: [Feature][Sort] Add Redis connector on flink 1.15
### Description

Parent issue: https://github.com/apache/inlong/issues/8336

### Use case

No response

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by XiaoYou201**
### Prepare a Pull Request

-  [INLONG-9835] [Feature][Sort] Add Redis connector on flink 1.15

- Fixes #8948

### Motivation

Due to the version of flink update to 1.15.4, the redis connector should be updated and tested.

### Modifications
According to the official website(https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sourcessinks/) introduce.The Metadata, Planning, and Runtime please refer to official website.

Accomplish customize source and sink about the redis-connector of flink 1.15.4. This work includes three main parts as follow.

H7EYD7D`P`~8D@DH20{VBQ7

#### 1. RedisDynamicTableFactory
This part mainly override two methods.
- createDynamicTableSource()

This method firtst validate the support commands and then return a RedisDynamicTableSource object.
- createDynamicTableSink()

This method firtst validate the support commands and then return a RedisDynamicTableSink object.
#### 2. RedisDynamicTableSource And RedisDynamicTableSink
- RedisDynamicTableSource

This class validate the config properties and  pass those properties and RedisRowDataLookupFunction object into TableFunctionProvider.of() method to  generate a LookupRuntimeProvider object which get data from redis servers.
- RedisDynamicTableSink

This class validate the config properties and  pass those properties and RedisSinkFunction object into SinkFunctionProvider.of() method to  generate SinkRuntimeProvider object which sink data to redis servers.
#### 3. RedisLookupFunction And RedisSinkFunction
- RedisLookupFunction

This class has a method called eval(), the eval() include the logical that get data from redis servers.Now, it has four request types to get different data (GET, HGET, ZREVRANK, ZSCORE).Due to flexible interface design, The request type can be quickly expand.
- RedisSinkFunction

This class has a method called flushInternal(), include the logical that sink data to redis servers.Now, it has three request types to sink different data (PLAIN, HASH, BITMAP).
### Documentation

  - Does this pull request introduce a new feature? (yes)

### CODE REVIEW

**Comment by dockerzhang** [comment]
@XiaoYou201 please add more details about this PR.

**Comment by EMsnap** [comment]
Please fix your uts and build

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/sink/RedisDynamicTableSink.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                        auditHostAndPorts);
+                break;
+            default:
+                throw new UnsupportedOperationException();
+        }
+
+        return SinkFunctionProvider.of(redisSinkFunction);
+    }
+
+    private int getValueStartIndexInSchema(
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/sink/RedisDynamicTableSink.java:160]: method not used

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/pom.xml [hunk 2]
```diff
@@ -43,6 +43,8 @@
         <module>hbase</module>
         <module>hudi</module>
         <module>kafka</module>
+        <module>redis</module>
+
```
Comment (aloyszhang) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/pom.xml:47]:
```suggestion
        <module>redis</module>
```

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/pom.xml [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-shaded-jackson</artifactId>
+            <version>2.12.4-15.0</version>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.apache.flink/flink-table-api-java-bridge -->
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-table-api-java-bridge</artifactId>
+            <version>1.15.4</version>
```
Comment (aloyszhang) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/pom.xml:47]: replace with ${flink.version}

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/config/RedisLookupOptions.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    public long getCacheExpireMs() {
+        return cacheExpireMs;
+    }
+
+    public int getMaxRetryTimes() {
+        return maxRetryTimes;
+    }
+
+    public boolean getLookupAsync() {
```
Comment (aloyszhang) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/config/RedisLookupOptions.java:66]: Not used, plz remove

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/handler/InlongJedisConfigHandler.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisConfigBase;
+import org.apache.flink.streaming.connectors.redis.common.hanlder.RedisHandler;
+
+/**
+ * handler to create flink jedis config.
+ */
+public interface InlongJedisConfigHandler extends RedisHandler {
+
+    /**
+     * create flink jedis config use sepecified properties.
```
Comment (aloyszhang) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/handler/InlongJedisConfigHandler.java:30]:
```suggestion
     * create flink jedis config use specified properties.
```

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/config/RedisDataType.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    SET,
+
+    /**
+     * Bitmaps are not an actual data type, but a set of bit-oriented operations defined on the String type. <br/>
+     * Since strings are binary safe blobs and their maximum length is 512 MB, <br/>
+     * they are suitable to set up to 2^32 different bits.
+     */
+    BITMAP,
+    ;
```
[the remark this answers is not in the document]
Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/config/RedisDataType.java]: thanks suggestion, fixed

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/pom.xml (hunk 2 above)
Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/pom.xml]: thanks suggestion, fixed

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/pom.xml (hunk 3 above)
Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/pom.xml]: thanks suggestion, fixed

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/handler/InlongJedisConfigHandler.java (hunk 5 above)
Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/main/java/org/apache/inlong/sort/redis/common/handler/InlongJedisConfigHandler.java]: thanks suggestion, fixed

**Review by aloyszhang — APPROVED** [review]

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/test/java/org/apache/inlong/sort/redis/RedisTableTest.java [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+
+        tableEnv.executeSql(dim);
+        tableEnv.executeSql(source);
+        tableEnv.executeSql(sink);
+        String sql =
+                " insert into sink"
+                        + " select '1_1',d.aaa, d.bbb from source s"
+                        + "  left join dim for system_time as of s.proctime as d "
+                        + " on d.aaa = s.aaa and d.bbb = s.bbb";
+        String testSql = " select concat_ws('_', s.aaa, s.aaa),d.bbb from source s"
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/test/java/org/apache/inlong/sort/redis/RedisTableTest.java:189]: testsql is not test please add a test case for it

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/test/java/org/apache/inlong/sort/redis/RedisTableTest.java [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+                        + "  left join dim for system_time as of s.proctime as d "
+                        + " on d.aaa = s.aaa";
+
+        tableEnv.executeSql(sql);
+
+        await().atMost(20, TimeUnit.SECONDS).untilAsserted(() -> {
+            assertEquals("value_1_1", jedis.get("1_1"));
+            assertEquals("value_2_2", jedis.get("2_2"));
+        });
+        //
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/test/java/org/apache/inlong/sort/redis/RedisTableTest.java:127]: remove this please

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/test/java/org/apache/inlong/sort/redis/RedisTableTest.java [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+public class RedisTableTest {
+
+    private static int redisPort;
+
+    private static RedisServer redisServer;
+
+    @BeforeClass
+    public static void setup() {
+        redisPort = NetUtils.getAvailablePort().getPort();
+        // redisPort = 6379;
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/redis/src/test/java/org/apache/inlong/sort/redis/RedisTableTest.java:47]: remove useless code

**Review by EMsnap — COMMENTED** [review]
should put declaration for copied codes under licenses/inlong-sort-connectors/LICENSE , thanks

**Review by EMsnap — COMMENTED** [review]
add dependency in inlong-sort/sort-core/pom.xml under flink v1.15 , and  also add in src/main/assemblies/sort-connectors-v1.15.xml

**Review by EMsnap — APPROVED** [review]

**Review by aloyszhang — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

**Pull request merged**